README.md
# MotorCortex-Player

**Table of Contents**

- [MotorCortex-Player](#motorcortex-player)
  - [Demo](#demo)
- [Intro / Features](#intro--features)
- [Getting Started](#getting-started)
  - [Installation](#installation)
  - [Importing](#importing)
- [Creating Player](#creating-player)
  - [Options](#options)
  - [Methods](#methods)
  - [Themes](#themes)
    - [transparent](#transparent)
    - [blue](#blue)
    - [green](#green)
    - [yellow](#yellow)
    - [dark](#dark)
    - [darkGold](#darkgold)
    - [whiteGold](#whitegold)
    - [custom](#custom)
  - [DonkeyClip QueryParams](#donkeyclip-queryparams)
- [Contributing](#contributing)
- [License](#license)
- [Sponsored by](#sponsored-by)

## Demo

[Check it out here](https://donkeyclip.github.io/motorcortex-player/demo/)

# Intro / Features

MotorCortex Player is a video-like player specially designed for MotorCortex Clips.
Among other functionality, the player allows the user to play/pause and seek a Clip, control the master volume and the playback speed and more.

What does our player include?

![Player Controls Image](https://donkey-spaces.ams3.digitaloceanspaces.com/assets/motorcortex-player/teo-player-numbers.png?)

1. Play/pause button
2. Master volume
3. Time: the time passed on the Clip's playback/total clip time
4. Loop: Click on this button if you want to set a loop on Clip's playback. Two grey anchor points will appear on the timeline from where you can define the loop boundaries.
5. Options: Click to open the options panel
6. Full Screen: Clip to go fullscreen
7. Pointer Events: Enabling "pointer-events" will allow your mouse to access the contents of the Clip. Otherwise, all mouse (pointer) events are disabled on the Clip's content
8. Show volume: shows or hides the master volume control (#2)
9. Speed: click to open the speed change control
10. Player's background-color
11. The actual motorcortex clip

# Getting Started

## Installation

```bash
$ npm install --save @donkeyclip/motorcortex-player
# OR
$ yarn add @donkeyclip/motorcortex-player
```

## Importing

```javascript
import MCPlayer from "@donkeyclip/motorcortex-player";
```

# Creating Player

The player can be used with any MotorCortex Clip. Here is an example with the minimum possible configuration:

```javascript
const player = new MCPlayer({ clip: MyClip });
```

## Options

Except for the (mandatory) `clip` option that should be passed on the player on instantiation there is a list of other supported options:
| Option | Default value | Description |
| ------------- |:-------------:| -----|
| clip | - | The MotorCortex Clip object that the Player will control |
| showVolume | false | Show or hide the master volume control (#2) |
| host | null | The Player by default "wraps" the Clip and seats at the bottom of its container taking advantage of the full width of it. In case you want the Player to get rendered on a different host (element) you can pass this element here |
| buttons | { loop: true, fullScreen: true, settings: true }| An object via which you can select which of the three buttons (#4: Loop via the key "loop", #5: Settings via the key "settings", #6: Full Screen via the key "fullScreen"). Enable and disable the buttons by setting their corresponding keys to true or false, accordingly |
| timeFormat | ss | This option allows you to set the time units on the Time (#3) of the Player. The default is ss, which means the Time will be presented in the form: < elapsed seconds >/< total seconds >. In the case you want a representation in the form: ms/ms you can set this option to "ms" |
| theme | "transparent" | Here you can pick one of the available themes of the player. The available themes are: "transparent", "blue", "yellow", "green","whiteGold","darkGold","dark" and "default" |
| themeCSS| - | Define your own theme using this property. [Checkout how to define custom themes](#custom)
| volume| 1 | With this option you can define the volume value. Acceptable values are any number between or equal to 0 and 1 |
| speed| 1 | A property from witch you can select the execution speed. Any value positive or negative is acceptable|
| loop|false| A boolean indicator whether the clip will initalize with the loop option on|
| autoPlay|false| With this option you can auto-start the clip when the loading is complete|
| controls|true| This option allows you to select if the player controls will be visible, therefore availble to the users. You cannot play or pause the clip by click events when this option is disabled. Scroll events are available if the player's type is set to "scroller" |
| scaleToFit|true| A property to define if the clip will be auto-scaled to fit to its' host element or not|
| overflow| "hidden" | A property to define if the you want elements outside the clip borders to be visible or not. Any css overflow values are acceptable|
| outline|unset| A property to define the border style of your clip. This can be useful on edit time. Any css border value is accepted|
| visible|unset| A property to define if you want your clip to be above the player controls. Accepted values are undefined, "always"|
| pointerEvents|true| This options adds a transparent div element on top of the clip in order to prevent user events (eg selection) on the clip element. When enabled, any click event on the clip triggers the play or pause event|
| backgroundColor|'black'| A css color value to define the players background |
| speedValues| [ -1, 0.5, 1, 2] | An array of number values to provide as supported options on the speed setting panel. This option will not prevent the user to change the clip speed programmatically to another value not provided in the speedValues property|
| muted| false | A Boolean property to define if the clip will be muted on start|
| currentScript| <script></script>| (Interests donkeyclip developers) If the clip will be loaded from a script with an external source where the player will be included, you can provide the script element refernce and use all the available control properties as data-\* attributes of the script (eg currentScript=document.currentScript)|
| millisecond| 0| The millisecond from witch the clip will start|
| thumbnail| - | A thumbnail image url to display in idle state|
| thumbnailColor| - | A thumbnail background color to display in idle state|
| type| "default" | You can select of the types "scroller" or "default". If you don't define this property the default player is active|
|scrollAnimation|false| If you set this property to true you can add animation to you scroll by extending the time of scroll|
|maxScrollStorage|50| You can store scroll events to extend the scroll experience. This property works best with scrollAnimation set to true and it takes an interger value greater than zero|
|sections|-| You can define sections as an array of milliseconds when you want to ankor the scroll events. (eg [0,1000,2000,3000]). Note that the zero(start) and clip's duration(end) are needed as values.

## Methods

| Name                |                Arguments                 | Description                                                                                                                                                                                                                                                               |
| ------------------- | :--------------------------------------: | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| onMillisecondChange |          function(millisecond)           | A function callback to be called after every millisecond change of the clip                                                                                                                                                                                               |
| goToMillisecond     | function(millisecond, clipCommands = {}) | A function from witch you can create a journey over the clips timeline. The clipCommands accepts the properties 'before' and 'after', from where you can for example pause the clip before the journey and 'play' it after (eg clipCommands={before:'pause',after:'play'} |
| createLoop          |         function(msStart, msEnd)         | With this method your can define a loop with boundaries. The loop option must be set to true for this to work                                                                                                                                                             |
| changeSettings      |           function(newOptions)           | This function allows you to manipulate the player options after the initialization. All the options provided in the 'Options' section can be used here                                                                                                                    |
| changeInitParams    |       the clips initparams object        | You can change any initial parameters of the clip by providing new values through this method. Any parameters that won't be defined here will remain the same as the previous state of initial parameters.                                                                |
| enterFullScreen     |                    -                     | You can enter fullscreen mode with code                                                                                                                                                                                                                                   |
| exitFullScreen      |                    -                     | You can exit fullscreen mode with code                                                                                                                                                                                                                                    |

```javascript
//example using player methods
import MClayer from "@donkeyclip/motorcortex-player";
const player = new MCPlayer({
  clip: MyClip,
  onMillisecondChange: (ms) =>
    console.log("There goes another millisecond. I am gonna live forever"),
});
player.changeSettings({ controls: false, speed: "2", loop: true });
player.createLoop(1000, 3000);
player.changeInitParams({ text: "hello world" });
```

## Themes

### transparent

![Player Controls](https://donkey-spaces.ams3.digitaloceanspaces.com/assets/motorcortex-player/transparent.png?)

### blue

![Player Controls](https://donkey-spaces.ams3.digitaloceanspaces.com/assets/motorcortex-player/blue.png?)

### green

![Player Controls](https://donkey-spaces.ams3.digitaloceanspaces.com/assets/motorcortex-player/green.png?)

### yellow

![Player Controls](https://donkey-spaces.ams3.digitaloceanspaces.com/assets/motorcortex-player/yellow.png?)

### dark

![Player Controls](https://donkey-spaces.ams3.cdn.digitaloceanspaces.com/assets/motorcortex-player/dark.png?)

### darkGold

![Player Controls](https://donkey-spaces.ams3.digitaloceanspaces.com/assets/motorcortex-player/darkGold.png?)

### whiteGold

![Player Controls](https://donkey-spaces.ams3.digitaloceanspaces.com/assets/motorcortex-player/whiteGold.png?)

### custom

All you have to do is define the following css text within the property `themeCSS` and use the name in the property theme of player options

```javascript
const player = new MCPlayer({
  clip: MyClip,
  themeCSS: `
  .--mc-player.theme-test {
    --activeColor: 136, 136, 500;
    --defaultColor: 136, 136, 136;
    --backgroundColor: 29, 31, 37, 1;
    --loopBarColor: rgba(var(--activeColor), 0.2);
    --grad-display: none;
  }`,
  theme: "theme-test",
});
```

## DonkeyClip QueryParams

All player options can be exploited by donkeyclips through query parameters. Any option of type boolean can be enabled without the need of the value "true". To disable options that are enabled by default you have to set the value the "false" for example:

```
https://api.donkeyclip.com/v1/embed/xxxxxxxxxx?controls=false&autoPlay&scaleToFit&theme=yellow&speed=2
```

For development purposes you can use the following query parameters to test different MotorCortex, Player, and Bundler versions:

- mcversion
- playerversion
- bundlerversion

For example: `https://api.donkeyclip.com/v1/embed/xxxxxxxxxx?controls&mcversion=7.4.0&playerversion=2.5.0&bundlerversion=7.0.1`

# Contributing

In general, we follow the "fork-and-pull" Git workflow, so if you want to submit patches and additions you should follow the next steps:

1. **Fork** the repo on GitHub
2. **Clone** the project to your own machine
3. **Commit** changes to your own branch
4. **Push** your work back up to your fork
5. Submit a **Pull request** so that we can review your changes

# License

[MIT License](https://opensource.org/licenses/MIT)

# Sponsored by

[<img src="https://presskit.donkeyclip.com/logos/donkey%20clip%20logo.svg" width=250></img>](https://donkeyclip.com)


CHANGELOG.md
# Changelog

## [2.10.11](https://github.com/donkeyclip/motorcortex-player/compare/v2.10.10...v2.10.11) (2023-09-01)


### Bug Fixes

* esm build and spread issues ([#240](https://github.com/donkeyclip/motorcortex-player/issues/240)) ([e845453](https://github.com/donkeyclip/motorcortex-player/commit/e8454538c5ebc4d21b28236770ccbbd3e94342e7))

## [2.10.10](https://github.com/donkeyclip/motorcortex-player/compare/v2.10.9...v2.10.10) (2023-08-28)

### Bug Fixes

- new esm.min files for build ([b4a339d](https://github.com/donkeyclip/motorcortex-player/commit/b4a339dc256055e23a54ba96c161745232f5096d))

## [2.10.9](https://github.com/donkeyclip/motorcortex-player/compare/v2.10.8...v2.10.9) (2023-07-14)

### Bug Fixes

- add host document functionality to incude mc clips in iframe ([7634afe](https://github.com/donkeyclip/motorcortex-player/commit/7634afeaf96c63aacb0c8d036e936be6afda52cb))
- Typo fixes by misspell-fixer ([b8eb97f](https://github.com/donkeyclip/motorcortex-player/commit/b8eb97fa90cec7cfb029733f2fe481ae6a45cdbc))

## [2.10.8](https://github.com/donkeyclip/motorcortex-player/compare/v2.10.7...v2.10.8) (2023-03-22)

### Bug Fixes

- player should keep the millisecond on change initparams, also the runtime state ([c88a9da](https://github.com/donkeyclip/motorcortex-player/commit/c88a9daabf2943d7be4f11c478dece7e2889f46e))
- player should keep the millisecond on change initparams, also the runtime state ([d7a5607](https://github.com/donkeyclip/motorcortex-player/commit/d7a5607990fce61ff187a6d6bd17af4ed46070bc))
- when scale event occurred the player did not consider the case when scaletofit was cover ([bd29f67](https://github.com/donkeyclip/motorcortex-player/commit/bd29f67eedeca3f579b3128e55bb3c3399cc073c))

## [2.10.7](https://github.com/donkeyclip/motorcortex-player/compare/v2.10.6...v2.10.7) (2022-10-22)

### Bug Fixes

- change the total time on change initial parameters ([b457215](https://github.com/donkeyclip/motorcortex-player/commit/b45721548f34981c3dbedacd6f4acde07cfd17dc))
- mc themes and loopbar ([44e9bee](https://github.com/donkeyclip/motorcortex-player/commit/44e9bee54e7bf92e9ced62d95053d1dc26b88ab2))
- open dc's home if no clip id ([33e4150](https://github.com/donkeyclip/motorcortex-player/commit/33e41508db5fa201a4be8602663ba089697b5481))

## [2.10.6](https://github.com/donkeyclip/motorcortex-player/compare/v2.10.5...v2.10.6) (2022-10-11)

### Bug Fixes

- player should check if the user is from mobile and then prevent default on progressbar move ([192f6f7](https://github.com/donkeyclip/motorcortex-player/commit/192f6f792a7636631ad4cc28c97d404ef38577fe))

## [2.10.5](https://github.com/donkeyclip/motorcortex-player/compare/v2.10.4...v2.10.5) (2022-06-21)

### Bug Fixes

- correctly handled undefined or empty init params ([d687d70](https://github.com/donkeyclip/motorcortex-player/commit/d687d70cf1913dc673325169df7b949d4347bbe7))

## [2.10.4](https://github.com/donkeyclip/motorcortex-player/compare/v2.10.3...v2.10.4) (2022-06-20)

### Bug Fixes

- access parent window in try catch ([3533bf8](https://github.com/donkeyclip/motorcortex-player/commit/3533bf8d3e40204aea9f1ec4914d67bc30e959eb))

## [2.10.3](https://github.com/donkeyclip/motorcortex-player/compare/v2.10.2...v2.10.3) (2022-06-10)

### Bug Fixes

- show donkeyclip button ([67b3f98](https://github.com/donkeyclip/motorcortex-player/commit/67b3f98cefd10ff00cc1652331710b442b4ca61c))

## [2.10.2](https://github.com/donkeyclip/motorcortex-player/compare/v2.10.1...v2.10.2) (2022-06-07)

### Bug Fixes

- add enter and exit fullscreen functions, prevent scroll on progress bar move ([b573b49](https://github.com/donkeyclip/motorcortex-player/commit/b573b49d6ce6a212e6610b62527e6601c28dc0e9))
- update readme ([952754d](https://github.com/donkeyclip/motorcortex-player/commit/952754d4e7a0a4d17dac38ced8a0fe81d000be45))

### [2.10.1](https://github.com/donkeyclip/motorcortex-player/compare/v2.10.0...v2.10.1) (2022-05-10)

### Bug Fixes

- less ifs ([3207fe9](https://github.com/donkeyclip/motorcortex-player/commit/3207fe9e65c6baf9312d27ba6ea47c7679cc8265))
- remove typo ([326864b](https://github.com/donkeyclip/motorcortex-player/commit/326864bf901fd1f99a1d697fafc554b4739f8cbe))
- show dcom button ([c52d7b8](https://github.com/donkeyclip/motorcortex-player/commit/c52d7b89d233185d32d350c28e4b5e35394570d9))

## [2.10.0](https://github.com/donkeyclip/motorcortex-player/compare/v2.9.7...v2.10.0) (2022-05-09)

### Features

- add donkeyclip button in all pages except dccom ([33c630d](https://github.com/donkeyclip/motorcortex-player/commit/33c630d73b38de52441b544904d08642a97027c5))

### [2.9.7](https://github.com/donkeyclip/motorcortex-player/compare/v2.9.6...v2.9.7) (2022-05-08)

### Bug Fixes

- loopbar handlers ([1d91607](https://github.com/donkeyclip/motorcortex-player/commit/1d916072a3389e4085d7aac0b28fcd32bf4c1214))
- return result values of change init params ([f835943](https://github.com/donkeyclip/motorcortex-player/commit/f8359437232faae3245b425612e218482792cb0d))

### [2.9.6](https://github.com/donkeyclip/motorcortex-player/compare/v2.9.5...v2.9.6) (2022-04-30)

### Bug Fixes

- add spinner and fix clip error bug ([63526cd](https://github.com/donkeyclip/motorcortex-player/commit/63526cd18efc4a075b67359e0eab07eb4b75919f))

### [2.9.5](https://github.com/donkeyclip/motorcortex-player/compare/v2.9.4...v2.9.5) (2022-04-20)

### Bug Fixes

- progressbar dims ([5fce4b8](https://github.com/donkeyclip/motorcortex-player/commit/5fce4b85ff1f77510a061f8f016d5662a385fcfb))

### [2.9.4](https://github.com/donkeyclip/motorcortex-player/compare/v2.9.3...v2.9.4) (2022-04-19)

### Bug Fixes

- user-select autoprefixes and player ([3695c2b](https://github.com/donkeyclip/motorcortex-player/commit/3695c2b52e7171940f0c2bba5a95743f6c8eddbc))

### [2.9.3](https://github.com/donkeyclip/motorcortex-player/compare/v2.9.2...v2.9.3) (2022-04-15)

### Bug Fixes


src/style.js
module.exports = theme => `
.background {
  background-color: black;
  width:100%;
  height:100%;
  position:absolute;
  top:0px;
  left:0px;
  z-index:-2000;
}
.full-screen #mc-player-controls {
  position:fixed;
  left:0px;
  bottom:0px;
}

.full-screen #mc-player-settings-panel {
  position:fixed;
  bottom: 45px;
}

.svg {
  fill: ${theme["svg-color"]};
  stroke: ${theme["svg-color"]};
}

.svg-selected svg{
  fill: ${theme["svg-selected-color"]};
  stroke: ${theme["svg-selected-color"]};
}
#mc-player-hover-display{
    border: ${theme["preview-border"]};
    max-width:300px;
    display: flex;
    overflow:hidden;
    background-color: black;
    position: absolute;
    bottom: 14px;
    left: 0px;
    align-items: flex-end;
    justify-content: center;
}

#mc-player-hover-millisecond {
  background-color: ${theme["hms-background-color"]};
  padding:3px;
  height:18px;
  margin:0px;
  line-height:12px;
  font-size:10px;
  text-align: center;
  min-width:20px;
  max-width:100px;
  z-index:2;
}
#mc-player,
#mc-player ::before,
#mc-player ::after,
#mc-player div,
#mc-player p,
#mc-player span,
#mc-player ul,
#mc-player li {
  font-weight: 400;
  line-height: 1.9 !important;
  color: ${theme["color"]};
  font-family: "Century Gothic", CenturyGothic, AppleGothic, sans-serif;
  box-sizing:border-box;
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}
#mc-player {
  line-height: 1.9;
  font-size: 12px;
  overflow:hidden;
  height: calc(100% + ${theme["controls-position"]});
  width:100%;
  position: absolute;
  top: 0px;
  left: 0px;
  
}

#mc-player-settings-speed-hide {
  text-align:right;
}

.grad {
  background-image: linear-gradient(
    rgba(100,100,100,00.01),
    rgba(100,100,100,00.02),
    rgba(100,100,100,00.03),
    rgba(100,100,100,0.04),
    rgba(100,100,100,0.05),
    rgba(0,0,0,0.06),
    rgba(0,0,0,0.07),
    rgba(0,0,0,0.08),
    rgba(0,0,0,0.09),
    rgba(0,0,0,0.1),
    rgba(0,0,0,0.2),
    rgba(0,0,0,0.3),
    rgba(0,0,0,0.4),
    rgba(0,0,0,0.4),
    rgba(0,0,0,0.5),
    rgba(0,0,0,0.6),
    rgba(0,0,0,0.7),
    rgba(0,0,0,0.8),
    rgba(0,0,0,0.9),
    rgba(0,0,0,1)
  );
  position:absolute;
  width:100%;
  height:${theme["grad-height"]};
  left:0px;
  bottom:0px;
}

#mc-player-controls {
  background-color: ${theme["background-color"]};
  border: ${theme["controls-border"]};
  position: absolute;
  bottom: ${theme["controls-bottom"]};
  left: 0px;
  width: 100%;
  height: 40px;
}

#mc-player-totalbar {
  width: calc(100% - 20px);
  height: 5px;
  margin: 0px 10px 0px 10px;
  background-color: #505056;
  position: relative;
  top: 0px;
  left: 0px;
}

#mc-player-loopbar {
  position: absolute;
  height: 100%;
  width: 100%;
  top: 0px;
  left: 0px;
  background-color: ${theme["loopbar-color"]};
}

.mc-player-loop-boundaries {
  transform:translate(-50%,-37%);
  position:absolute;
  width:18px;
  background-color:${theme["loopbar-color"]};
  height:18px;
  border-radius:10px;
  z-index:40;
}

#mc-player-helperbar {
  position: absolute;
  height: 20px;
  top: -7px;
  left: 0px;
  right: 0px;
  z-index:2;
}

#mc-player-runningbar {
  position: relative;
  width: 0px;
  max-width:100%;
  height: 100%;
  background-color: ${theme["runningbar-color"]};
}

#mc-player-cursor {
  transform:translate(50%,-36%);
  right: 0px;
  top: 0px;
  width: 0px;
  height: 0px;
  position: absolute;
  background-color: ${theme["cursor-color"]};
  border-radius: 10px;
  z-index: 5;
}

#mc-player-left-controls {
  display:inline-block;
  width:200px;
  height:35px;
}
#mc-player-time-display {
  display: table;
  text-align: center;
  width: auto;
  height: 34px;
  position: absolute;
  left: 90px;
  -webkit-transition: all 0.1s ease;
  -moz-transition: all 0.1s ease;
  transition: all 0.1s ease;
}

#mc-player-time-display span {
  display: table-cell;
  vertical-align: middle;
}

#mc-player-status-btn {
  opacity: ${theme["button-opacity"]};
  background-repeat: no-repeat;
  background-size: 100% 100%;
  width: 40px;
  height: 25px;
  position: absolute;
  overflow:visible;
  left: 0px;
  bottom: 0px;
  margin: 10px 5px 5px 5px;
  text-align: center;
  z-index:3;
}

#mc-player-volume {
  opacity: ${theme["button-opacity"]};
  background-repeat: no-repeat;
  background-size: 100% 100%;
  position: absolute;
  left: 45px;
  bottom: 5px;
  margin: 10px 5px 5px 5px;
  text-align: center;
  z-index:3;
}
#mc-player-volume-btn {
  width: 20px;
  height: 15px;
}

#mc-player-volumebar {
  width: 0px;
  height: 3px;
  background-color: ${theme["loopbar-color"]};
  position:absolute;
  left:25px;
  bottom:6px;
  -webkit-transition: all 0.1s ease;
  -moz-transition: all 0.1s ease;
  transition: all 0.1s ease;
}

#mc-player-volumebar-helper {
  position: absolute;
  width: 0px;
  height: 15px;
  left:25px;
  bottom:0px;
  z-index:10;
}

#mc-player-volumebar-active {
  position: relative;
  width: 0%;
  height: 100%;
  background-color: ${theme["color"]};
  position:relative;
  left:0px;
  bottom:0px;
}

#mc-player-volume-cursor {
  transform:translate(50%,-36%);
  right: 0px;
  top: 0px;
  width: 0px;
  height: 0px;
  position: absolute;
  background-color: ${theme["color"]};
  border-radius: 10px;
  z-index: 5;
}

.mc-player-loopbar-time {
  width:auto;
  height:12px;
  background-color:${theme["background-color"]};
  line-height:10px;
  font-size:10px;
}

#mc-player-loop-time { 
  position:absolute;
  right:85px;
  bottom:5px;
}

#mc-player-loop-btn {
  opacity: ${theme["button-opacity"]};
  background-repeat: no-repeat;
  background-size: 100% 100%;
  width: 15px;
  height: 15px;
  position: absolute;
  right: 55px;
  bottom: 5px;
  margin: 10px 5px 5px 5px;
}

#mc-player-settings-btn {
  opacity: ${theme["button-opacity"]};
  background-repeat: no-repeat;
  background-size: 100% 100%;
  width: 15px;
  height: 15px;
  position: absolute;
  right: 30px;
  bottom: 5px;
  margin: 10px 5px 5px 5px;
}

#mc-player-full-screen-btn {
  opacity: ${theme["button-opacity"]};
  background-repeat: no-repeat;
  background-size: 100% 100%;
  width: 15px;
  height: 15px;
  position: absolute;
  right: 5px;
  bottom: 5px;
  margin: 10px 5px 5px 5px;
}

.mc-player-speed-btn {
  position: relative;
  opacity: ${theme["button-opacity"]};
  width: 10px;
  height: 10px;
  display: inline-block;
}

#mc-player-settings-panel {
  box-sizing: border-box;
  position: absolute;
  background-color: ${theme["settings-background-color"]};
  bottom: ${theme["settings-panel-bottom"]};
  border: ${theme["border"]};
  right: 5px;
  width: 164px;
  height: 143px;
  padding: 5px;
  margin: 0px;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

#mc-player-speed-value-bar {
  position: relative;
  width: 5px;
  background-color: ${theme["speedbar-color"]};
  display: inline-block;
  box-sizing: border-box;
  height: 144px;
  float: left;
  margin-right:15px;
}

#mc-player-speed-value-helperbar {
  position: absolute;
  width: 25px;
  height: 144px;
  float: left;
  left: -5px;
  z-index:10;
}


#mc-player-speed-value-bar:hover,
#mc-player-speed-value-helperbar {
  cursor: pointer;
}

#mc-player-volumebar:hover,
#mc-player-volumebar-helper:hover,
#mc-player-volume-btn:hover,
#mc-player-volumebar:active,
#mc-player-volumebar-helper:active,
#mc-player-volume-btn:active {
  cursor:pointer;
}

#mc-player-speed-cursor {
  position: absolute;
  background-color: ${theme["speedbar-cursor-color"]};
  top: 0px;
  left: 0px;
}

#mc-player-speed-cursor div {
  position: absolute;
  background-color: ${theme["speedbar-cursor-color"]};
  left: -2.5px;
  top: -4px;
  width: 10px;
  height: 10px;
  border-radius: 5px;
}

#mc-player-speed-cursor:hover {
  cursor: pointer;
}

.mc-player-speed-value-step {
  width: 15px;
  background-color: ${theme["speedbar-color"]};
  display: inline-block;
  box-sizing: border-box;
  height: 2px;
  margin-top: 7px;
  margin-bottom: 7px;
  float: left;
}

#mc-player-speed-value {
  display: inline-block;
  box-sizing: border-box;
  height: 144px;
  text-align: left;
}

.mc-player-speed-value {
  box-sizing: border-box;
  height: 16px;
  font-size: 12px;
}

#mc-player-indicator {
  font-size: 8px !important;
  position: relative;
  bottom: 15px;
  color: ${theme["color"]};
  white-space: nowrap;
}

#mc-player-settings-panel.mc-player-settings-speed-panel {
  overflow: hidden;
  width: 80px;
  height: 195px;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

#mc-player-settings-panel.mc-player-settings-speed-panel .mc-player-speed-btn {
  float: left;
}

.mc-player-settings-speed-panel ul:first-child {
  text-align: right;
}

#mc-player-speed-current {
  float: right;
  padding-right: 10px
}

#mc-player-settings-panel .mc-player-speed-btn {
  float: right;
}

#mc-player-settings-panel ul {
  width: 100%;
  margin: 0px;
  padding: 0px;
  overflow: hidden;
}

#mc-player-settings-panel.mc-player-settings-speed-panel ul li {
  min-width: 70px;
}

#mc-player-settings-panel ul li.no-hover:hover {
  background-color: transparent;
  cursor: default;
}

div.mc-player-speed-value:hover {
  background-color: ${theme["hover-color"]};
  cursor: pointer;
}

#mc-player-settings-panel ul li {
  position: relative;
  width: 100%;
  min-width: 154px;
  list-style-type: none;
  margin: 0px;
  padding: 5px;
}

#mc-player-settings-panel ul li label {
  margin: 0px;
}

.switch {
  position: relative;
  display: inline-block;
  width: 35px;
  height: 18px;
}

.switch input {
  display: none;
}

.settings-switch {
  float: right;
}

.settings-switch:after {
  clear: both;
}

.slider {
  position: absolute;
  cursor: pointer;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: ${theme["slider-off-color"]};
  -webkit-transition: .4s;
  transition: .4s;
}

.slider:before {
  position: absolute;
  content: "";
  height: 16px;
  width: 16px;
  left: 1px;
  bottom: 1px;
  background-color: white;
  -webkit-transition: .4s;
  transition: .4s;
}

input:checked+.slider {
  background-color: ${theme["slider-on-color"]};
}

input:focus+.slider {
  box-shadow: 0 0 1px ${theme["slider-on-color"]};
}

input:checked+.slider:before {
  -webkit-transform: translateX(16px);
  -ms-transform: translateX(16px);
  transform: translateX(16px);
}


/* Rounded sliders */

.slider.round {
  border-radius: 34px;
}

.slider.round:before {
  border-radius: 50%;
}


.m-fadeOut {
  visibility: hidden;
  opacity: 0;
  transition: visibility 0s linear 300ms, opacity 300ms;
}

.m-fadeIn {
  visibility: visible;
  opacity: 1;
  transition: visibility 0s linear 0s, opacity 300ms;
}

#mc-player-settings-panel ul li:hover {
  background-color: ${theme["hover-color"]};
  cursor: pointer;
}

#mc-player-settings-panel ul li label:hover {
  cursor: pointer;
}

#mc-player-loopbar:hover {
  cursor: pointer;
}

#mc-player-status-btn:hover {
  cursor: pointer;
}

#mc-player-controls:active #mc-player-cursor,
#mc-player-controls:hover #mc-player-cursor  {
  width: 16px;
  height: 16px;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

#mc-player-left-controls:active #mc-player-volume-cursor,
#mc-player-left-controls:hover #mc-player-volume-cursor {
  width: 12px;
  height: 12px;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

#mc-player-left-controls:active .mc-player-volume-width-transition,
#mc-player-left-controls:hover .mc-player-volume-width-transition
 {
  width: 50px;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

#mc-player-left-controls:active .mc-player-time-width-transition,
#mc-player-left-controls:hover .mc-player-time-width-transition {
  left: 140px;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

#mc-player-settings-speed:hover .mc-player-speed-btn {
  opacity: 1;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

#mc-player-status-btn:hover {
  opacity: 1;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

#mc-player-loop-btn:hover {
  cursor: pointer;
  opacity: 1;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  transition: all 0.3s ease;
}
#mc-player-settings-btn:hover {
  cursor: pointer;
  opacity: 1;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  transition: all 0.3s ease;
}

#mc-player-full-screen-btn:hover {
  cursor: pointer;
  opacity: 1;
  -webkit-transition: all 0.3s ease;
  -moz-transition: all 0.3s ease;
  transition: all 0.3s ease;
}`;
